src/cmd/decode.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+    cmd::extrinsics::{load_metadata, ContractMessageTranscoder},
+    util::decode_hex,
+    DEFAULT_KEY_COL_WIDTH,
+};
+use anyhow::{Context, Result};
+use colored::Colorize as _;
+
+#[derive(Debug, Clone, clap::Args)]
+#[clap(name = "decode", about = "Decode input_data for a contract")]
+pub struct DecodeCommand {
+    /// Type of data
+    #[clap(arg_enum, short, long)]
+    r#type: DataType,
+    /// The data to decode
+    #[clap(long)]
+    data: String,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, clap::ArgEnum)]
+enum DataType {
+    Event,
+    Message,
+}
+
+impl DecodeCommand {
+    pub fn run(&self) -> Result<()> {
```

Comment:
I'd like to see a basic test for decoding a message and an event. to prove that it works. As with your recently reported issues it is easy to introduce regressions.There are a few options here:

1. A full end to end test like https://github.com/paritytech/cargo-contract/blob/82157f46f01197f662715a3f0a81dd9d749686d3/src/cmd/build.rs#L1445 which invokes the process.
2. Invoking `run` directly with some setup https://github.com/paritytech/cargo-contract/blob/82157f46f01197f662715a3f0a81dd9d749686d3/src/cmd/build.rs#L1445
3. Factor out a function and pass in the contract metadata.